README
NAME
    Mail::GPG - Handling of GnuPG encrypted / signed mails

SYNOPSIS
      use Mail::GPG;

      my $mg = Mail::GPG->new;

      my %keys_id2mail = $mg->query_keyring (
        search => 'joern@zyn.de',
      );

      my $entity = MIME::Entity->build (
        From     => 'joern@zyn.de',
        Subject  => "Mail::GPG Testmail",
        Data     => [ "Hiho, a nice encrypted mail" ],
        Encoding => "quoted-printable",
        Charset  => "iso-8859-1",
      );

      my $encrypted_entity = $mg->mime_sign_encrypt (
        entity     => $entity,
        key_id     => $key_id,
        password   => 'topsecret',
        recipients => [ 'niceguy@zyn.de' ],
      );

      my $mail_text = $encrypted_entity->as_string;

      # and a lot more...

      $mg->mime_sign ( ... );
      $mg->mime_encrypt ( ... );
      $mg->mime_sign_encrypt ( ... );

      $mg->armor_sign ( ... );
      $mg->armor_encrypt ( ... );
      $mg->armor_sign_encrypt ( ... );

      $mg->parse ( ... );

      $mg->decrypt ( ... );
      $mg->verify ( ... );

      $mg->is_encrypted ( ... );
      $mg->is_signed ( ... );
      $mg->is_signed_quick ( ... );

      $mg->get_decrypt_key ( ... );
      $mg->get_key_trust ( ... );

DESCRIPTION
    This Perl modules handles all the details of encrypting and signing
    Mails using GnuPG according to RFC 3156 and RFC 2440, that is OpenPGP
    MIME and traditional armor signed/encrypted mails.

PREREQUISITES
      Perl              >= 5.00503
      MIME-tools        >= 5.419
      MIME::QuotedPrint >= 2.20  (part of MIME-Base64 distribution)
      GnuPG::Interface  >= 0.33  (optionally with shipped patch applied)

    Since version 1.0.7 the Event module is loaded when present. This gives
    a small performance boost when handling big entities. If you like to
    fallback to the old IO::Select based behaviour set either

      $Mail::GPG::SKIP_EVENT    = 1
      $ENV{MAIL_GPG_SKIP_EVENT} = 1

    before loading Mail::GPG.

INSTALLATION
    Before you install Mail::GPG you may want to apply the shipped
    GnuPG::Interface patch. It just fixes a warning which is throwed on any
    keyring inspection. This is a known problem and reported to the author,
    hopefully it will be fixed upstream soon:

      % tar xvfz GnuPG-Interface-0.33.tar.gz
      % cd GnuPG-Interface-0.33
      % patch -p1 ../Mail-GPG.x.xx/patches/GnuPG-Interface-0.33.tru-record-type.txt
      % perl Makefile.PL
      % make test
      % make install

    Then install Mail::GPG

      % cd ../Mail-GPG-x.xx
      % perl Makefile.PL
      % make test
      % make install

    Mail::GPG has a bunch of tests which will create a temporary gpg keyring
    to be able to do real encryption and stuff. You need to have gpg in your
    path for the tests to succeed, otherwise all useful tests will be
    skipped.

    Note that the test 04.big needs some time, on an Athlon 1800XP about 12
    seconds, so be patient ;)

KNOWN BUGS
    Currently none. Please report any bugs to the author: Joern Reder <joern
    AT zyn.de>.

AUTHOR
    Joern Reder <joern AT zyn.de>

CONTACT
    You can contact me by email. Please place the module name "Mail::GPG"
    somewhere in the subject, because I filter my mails that way. I'm a
    native German speaker, but you can contact me in english as well.

COPYRIGHT
    Copyright (C) 2004-2006 by Joern Reder, All Rights Reserved.

    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
$Id: Changes,v 1.20 2011-08-03 08:53:29 joern Exp $

Revision history and release notes for Mail::GPG.

1.0.7 Wed Aug 3, 2011, joern
    Features:
    - Added support for empty passphrases. Thanks for the patch
      to hlein AT korelogic DOT com.
    - Improved gpg I/O performance and prevent rare blocking
      cases by using the Event module (if installed).
      
      Note: you can force the old IO::Select based algorithm
            by setting $Mail::GPG::SKIP_EVENT = 1 or
            $ENV{MAIL_GPG_SKIP_EVENT} = 1 before loading
            Mail::GPG.

1.0.6 Sat Nov 18, 2006, joern
    Bugfix:
    - Mail::GPG->parse() missed setting MIME::Parser->output_to_core(1)
      so temporary files were left. Thanks for the hint to
      David Morel <david.morel AT amakuru.net>. As well changed
      the test suite to make use of Mail::GPG->parse().

1.0.5 Sun Nov 12, 2006, joern
    Bugfixes:
    - Applied a patch from Alwin Stolk <pastolk AT xs4all.nl>
      which increases message decryption performance for big
      mail bodies.
    - Fixed POD typo. Thanks to Edward J. Sabol for the hint.

1.0.4 Fri Apr 14, 2006, joern
    Features:
    - Report signer's key fingerprint with get_sign_fingerprint()
      of Mail::GPG::Result. Requested by Peter Conrad <conrad AT
      tivano.de>.
    - Report multiple encryption recipients with the new methods
      get_enc_key_ids() and get_enc_mails() of Mail::GPG::Result.
    - Switch key id reporting from 32 to 64 bit length with
      the Mail::GPG attribute use_long_key_ids.
    - Suppress warnings from GnuPG::Interface about unknown recored
      type 'tru', making my patch at least in conjunction with
      Mail::GPG obsolete. The maintainer of GnuPG::Interface
      obviously isn't interested in new releases of his module :(.
    
    Bugfixes:
    - Don't rely on gnupg's exit code anywhere but inspect
      its status fd output only. Thanks for the hint to
      Peter Conrad.

1.0.3 Fri Dec 22, 2005, joern
    Notes:
    - No code changes, just removed MIME-tools patches from the
      distribution because the official MIME-tools 5.419 package
      now contains them.
    - That's why Mail::GPG instists on MIME-tools 5.419 from now on.
    - Removed all pointers to the MIME-tools patch from the
      documentation.

1.0.2 Sat Dec 17, 2005, joern
    Features:
    - Added a patch against MIME-tools 5.418, which hopefully
      gets into the official distribution soon. Thanks to
      Edward J. Sabol who helped me to get this patch right.

    Bugfixes:
    - Fixed a (potential) bug with Mail::GPG->get_key_trust(),
      which was reported and debugged by Edward J. Sabol, but
      couldn't be reproduced on my system. However the fix
      is harmless, so I added it.

1.0.1 Sun Dec 5, 2004, joern
    Features:
    - New methods requested by Kjetil Kjernsmo <kjetil AT kjernsmo.net>
      - Mail::GPG->is_signed_quick()
      - Mail::GPG->get_key_trust()
      - Mail::GPG::Result->get_sign_trust()
      - Mail::GPG::Result->get_enc_trust()
    - A patch for GnuPG::Interface 0.33 fixing a (harmless but
      annoying ;) warning about unknown record types on keyring
      inspection with newer GnuPG versions.

1.0.0 Sat Nov 20, 2004, joern
    Features:
    - New parse() convenience method for parsing a mail message.
      It uses MIME::Parser and distinguish between MIME and
      non-MIME messages, doing the right thing regarding reading
      decoded or encoded bodies.

    Bugfixes:
    - decrypt(): used encoded entity instead of decoded in case
      of a MIME message. Thanks for the report to Paul Murphy
      <pmurphy AT ionixpharma.com>.

0.98 Sun Jun 27, 2004, joern
    Bugfixes:
    - Mail::GPG->decrypt() died on encrypted+signed mails for
      which the public key was missing (and thus no verification
      possible). That's not what we expect, instead we can check
      the verification through the returned result object but
      have a successfully decrypted mail.

0.97 Tue Jun 15, 2004, joern
    Bugfixes:
    - Mail::GPG::Result->as_string() reported some warnings
      regarding uninitialized values. Thanks to Edward J. Sabol
      <sabol AT alderaan.gsfc.nasa.gov> for his patch.
    - mgpg-test script now has the -w switch

0.96 Sat May 29, 2004, joern
    Bugfixes:
    - Added missing dependency GnuPG::Interface in Makefile.PL
      Reported through cpan-testers.

0.95 Sat May 29, 2004, joern
    Features:
    - query_keyring() now returns all matching entries as a
      list of key-id/address pairs, which can be slurped
      into a hash. The old behaviour is still warranted, so
      this extension is backward compatible.
    - The signer's aliases are now reported with the new
      sign_mail_aliases attribute of Mail::GPG::Result. Thanks
      for the suggestion to Piotr Maj <ant AT kernelpanic.pl>.

    Bugfixes:
    - Some methods failed on systems with non-english
      locales. Thanks to Daniel Ciaglia <daniel AT ciaglia.de>
      and Piotr Maj for their reports.
    - Not a fix in Mail::GPG but in MIME-tools: some mail
      user agents (like mutt and sylpheed) add an empty line
      as a preamble before the first part of a MIME signed
      mail, if the message itself has attachments. MIME-tools
      suppress this line on output, so the signature gets
      invalid. Mail::GPG ships with a new patch to MIME-tools
      5.411 which fixes this problem. Thanks for the report
      to Daniel Ciaglia.

0.94 Sun Feb 15, 2004, joern
    Bugfixes:
    - New I/O multiplex code needed some adjustments to
      work with Perl 5.005_03, because its read() and seek()
      functions work on native filehandles only.

0.93 Sat Feb 14 2004, joern
    Bugfixes:
    - Perform multiplexed I/O with gpg to prevent buffering
      deadlocks which may occur with huge amounts of data.
      Added a new test t/04.big.t, which creates a 4 MB
      entity and signs it. This test takes some time (on

lib/Mail/GPG/Result.pm
package Mail::GPG::Result;

# $Id: Result.pm,v 1.9 2009-05-30 13:59:11 joern Exp $

use strict;

require Encode if $] >= 5.008;

sub get_mail_gpg		{ shift->{mail_gpg}			}

sub get_is_encrypted		{ shift->{is_encrypted}			}
sub get_enc_ok			{ shift->{enc_ok}			}
sub get_enc_key_id		{ shift->{enc_key_id}			}
sub get_enc_mail		{ shift->{enc_mail}			}
sub get_enc_key_ids             { shift->{enc_key_ids}                  }
sub get_enc_mails               { shift->{enc_mails}                    }
sub get_is_signed		{ shift->{is_signed}			}
sub get_sign_ok			{ shift->{sign_ok}			}
sub get_sign_state              { shift->{sign_state}                   }
sub get_sign_key_id		{ shift->{sign_key_id}			}
sub get_sign_mail		{ shift->{sign_mail}			}
sub get_sign_mail_aliases	{ shift->{sign_mail_aliases}		}
sub get_sign_fingerprint        { shift->{sign_fingerprint}             }

sub set_is_encrypted            { shift->{is_encrypted}         = $_[1] }
sub set_enc_ok                  { shift->{enc_ok}               = $_[1] }
sub set_enc_key_id              { shift->{enc_key_id}           = $_[1] }
sub set_enc_mail                { shift->{enc_mail}             = $_[1] }
sub set_enc_key_ids             { shift->{enc_key_ids}          = $_[1] }
sub set_enc_mails               { shift->{enc_mails}            = $_[1] }
sub set_is_signed               { shift->{is_signed}            = $_[1] }
sub set_sign_ok                 { shift->{sign_ok}              = $_[1] }
sub set_sign_state              { shift->{sign_state}           = $_[1] }
sub set_sign_key_id             { shift->{sign_key_id}          = $_[1] }
sub set_sign_mail               { shift->{sign_mail}            = $_[1] }
sub set_sign_mail_aliases       { shift->{sign_mail_aliases}    = $_[1] }
sub set_sign_fingerprint        { shift->{sign_fingerprint}     = $_[1] }

sub get_gpg_stdout		{ shift->{gpg_stdout}			}
sub get_gpg_stderr		{ shift->{gpg_stderr}			}
sub get_gpg_status              { shift->{gpg_status}                   }
sub get_gpg_rc			{ shift->{gpg_rc}			}

sub new {
    my $class = shift;
    my %par   = @_;
    my  ($mail_gpg, $gpg_stdout, $gpg_stderr, $gpg_status, $gpg_rc) =
    @par{'mail_gpg','gpg_stdout','gpg_stderr','gpg_status','gpg_rc'};

    #-- initialize reference attributes to prevent
    #-- dereferencing undef errors
    $gpg_stdout = \"" if not defined $gpg_stdout;
    $gpg_stderr = \"" if not defined $gpg_stderr;
    $gpg_status = \"" if not defined $gpg_status;

    my $self = bless {
        mail_gpg   => $mail_gpg,
        gpg_stdout => $gpg_stdout,
        gpg_stderr => $gpg_stderr,
        gpg_status => $gpg_status,
        gpg_rc     => $gpg_rc,
    }, $class;

    $self->analyze_result,

        return $self;
}

sub analyze_result {
    my $self = shift;

    my ($is_signed,         $sign_ok,      $sign_state,
        $sign_key_id,       $sign_mail,    $sign_fingerprint,
        @sign_mail_aliases, $is_encrypted, $enc_ok,
        @enc_key_ids,       @enc_mails
    );

    my $gpg_status = $self->get_gpg_status;
    my $gpg_stderr = $self->get_gpg_stderr;

    while ( $$gpg_status && $$gpg_status =~ m{^\[GNUPG:\]\s+(.*)$}mg ) {
        my $line = $1;
        if ( $line =~ /^(GOOD|EXP|EXPKEY|REVKEY|BAD)SIG\s+([^\s]+)\s+(.*)/ ) {
            my ( $state, $key_id, $mail ) = ( $1, $2, $3 );
            $is_signed   = 1;
            $sign_state  = $state;
            $sign_key_id = $key_id;
            $sign_mail   = decode($mail);
            $sign_ok     = $sign_state eq 'GOOD';
        }
        elsif ( $line =~ /^ERRSIG\s+([^\s]+)/ ) {
            $is_signed   = 1;
            $sign_key_id = $1;
        }
        elsif ( $line =~ /^VALIDSIG\s+([^\s]+)/ ) {
            $sign_fingerprint = $1;
        }
        elsif ( $line =~ /^ENC_TO\s+([^\s]+)/ ) {
            push @enc_key_ids, decode($1);
            $enc_mails[ @enc_key_ids - 1 ] = "";
        }
        elsif ( $line =~ /^USERID_HINT\s+([^\s]+)\s+(.*)/ ) {
            my ( $hint_key_id, $hint_text ) = ( $1, $2 );
            my $i = 0;
            foreach my $key_id (@enc_key_ids) {
                if ( $key_id eq $hint_key_id ) {
                    $enc_mails[$i] = decode($hint_text);
                }
                ++$i;
            }
        }
        elsif ( $line =~ /^BEGIN_DECRYPTION/ ) {
            $is_encrypted = 1;
        }
        elsif ( $line =~ /^DECRYPTION_OKAY/ ) {
            $enc_ok = 1;
        }
    }

    @sign_mail_aliases = $$gpg_stderr =~ /^gpg:\s+aka\s+"(.*?)"/mg;

    if ( !$self->get_mail_gpg->get_use_long_key_ids ) {
        for ( $sign_key_id, @enc_key_ids ) {
            $_ = substr( $_, -8, 8 ) if defined $_;
        }
    }

    $self->set_is_signed( $is_signed               || 0 );
    $self->set_sign_ok( $sign_ok                   || 0 );
    $self->set_sign_state( $sign_state             || "" );
    $self->set_sign_key_id( $sign_key_id           || "" );
    $self->set_sign_mail( $sign_mail               || "" );
    $self->set_sign_fingerprint( $sign_fingerprint || "" );
    $self->set_sign_mail_aliases( \@sign_mail_aliases );
    $self->set_is_encrypted( $is_encrypted || '0' );

    $self->set_enc_ok( $enc_ok || 0 );
    $self->set_enc_key_ids( \@enc_key_ids );
    $self->set_enc_mails( \@enc_mails );
    $self->set_enc_key_id( $enc_key_ids[0] || "" );
    $self->set_enc_mail( $enc_mails[0]     || "" );

    return $self;
}

sub decode {
    my ($str) = @_;
    return $str if not defined $str;
    $str =~ s/\\x(..)/chr(hex($1))/eg;
    Encode::_utf8_on($str) if $] >= 5.008;
    return $str;
}

sub as_string {
    my $self        = shift;
    my %par         = @_;
    my ($no_stdout) = $par{'no_stdout'};

    my ( $method, $string );
    foreach my $attr (
        qw (is_encrypted enc_ok enc_key_id enc_mail
        enc_key_ids enc_mails
        is_signed sign_ok sign_state sign_key_id
        sign_fingerprint sign_mail sign_mail_aliases
        sign_trust gpg_rc )
        ) {
        if ( $attr eq 'sign_mail_aliases' ) {
            foreach my $value ( @{ $self->get_sign_mail_aliases } ) {
                $string
                    .= sprintf( "%-18s: %s\n", "sign_mail_alias", $value );
            }
        }
        elsif ( $attr eq 'enc_key_ids' ) {
            foreach my $value ( @{ $self->get_enc_key_ids } ) {
                $string .= sprintf( "%-18s: %s\n", "enc_key_ids", $value );
            }
        }
        elsif ( $attr eq 'enc_mails' ) {
            foreach my $value ( @{ $self->get_enc_mails } ) {
                $string .= sprintf( "%-18s: %s\n", "enc_mails", $value );
            }
        }
        else {
            $method = "get_$attr";
            my $value = $self->$method;
            $value = "" unless defined $value;
            $string .= sprintf( "%-18s: %s\n", $attr, $value );
        }
    }

    my $stdout = ${ $self->get_gpg_stdout };
    my $stderr = ${ $self->get_gpg_stderr };
    my $status = ${ $self->get_gpg_status };

    for ( $stdout, $stderr, $status ) {
        next unless $_;
        s/\n/\n                    /g;
        s/\s+$//;
    }

    $string .= sprintf( "%-18s: %s\n", "gpg_stdout", $stdout || '' )
        if not $no_stdout;
    $string .= sprintf( "%-18s: %s\n", "gpg_stderr", $stderr || '' );
    $string .= sprintf( "%-18s: %s\n", "gpg_status", $status || '' );

    return $string;
}

sub as_short_string {
    my $self = shift;

    my $string;

    if ( $self->get_is_encrypted ) {
        $string .= "ENC("
            . $self->get_enc_mail . ", "
            . $self->get_enc_key_id . ", "
            . ( $self->get_enc_ok ? "OK" : "NOK" ) . ") - ";
    }
    else {
        $string .= "NOENC - ";
    }

    if ( $self->get_is_signed ) {
        $string .= "SIGN("
            . $self->get_sign_mail . ", "
            . $self->get_sign_key_id . ", "
            . $self->get_sign_state . ") - ";
    }
    else {
        $string .= "NOSIGN - ";
    }

    $string =~ s/ - $//;

    return $string;
}

sub get_sign_trust {
    my $self = shift;

    return $self->{sign_trust} if exists $self->{sign_trust};

    my $trust = $self->get_mail_gpg->get_key_trust(
        key_id => $self->get_sign_key_id );

    return $self->{sign_trust} = $trust;
}

1;

__END__


=head1 NAME

Mail::GPG::Result - Mail::GPG decryption and verification results

=head1 SYNOPSIS

  $result = $mg->verify (
    entity => $entity
  );
  
  ($decrypted_entity, $result) = $mg->decrypt (
    entity => $entity,
  );

  $long_string  = $result->as_string ( ... );
  $short_string = $result->as_short_string;

  $encrypted           = $result->get_is_encrypted;
  $decryption_ok       = $result->get_enc_ok;
  $encryption_key_id   = $result->get_enc_key_id;
  $encryption_mail     = $result->get_enc_mail;
  $enc_key_ids         = $result->get_enc_key_ids;
  $enc_mails           = $result->get_enc_mails;

  $signed              = $result->get_is_signed;
  $signature_ok        = $result->get_sign_ok;
  $signed_key          = $result->get_sign_key_id;
  $signed_fingerprint  = $result->get_sign_fingerprint;
  $trust               = $result->get_sign_trust;
  $signed_mail         = $result->get_sign_mail;
  $signed_mail_aliases = $result->get_sign_mail_aliases;

  $stdout_sref         = $result->get_gpg_stdout;
  $stderr_sref         = $result->get_gpg_stderr;
  $status_sref         = $result->get_gpg_status;
  $gpg_exit_code       = $result->get_gpg_rc;

=head1 DESCRIPTION

This class encapsulates decryption and verification results
of Mail::GPG. You never create objects of this class yourself,
they're all returned by Mail::GPG.

=head1 ATTRIBUTES

This class mainly has a bunch of attributes which reflect the
result of a Mail::GPG operation. You can read these attributes
with B<get>_attribute.

=over 4

=item B<is_encrypted>

Indicates whether an entity was encrypted or not.

=item B<enc_ok>

Indicates whether decryption of an entity was successful or not.

=item B<enc_key_id>

The recipient's key id of an encrypted mail. This is the first
key reported by gnupg in case the mail has more than one
recipient.

=item B<enc_mail>

The recipient's mail address of an encrypted mail. This is the first
mail address reported by gnupg in case the mail has more than one
recipient.

=item B<enc_key_ids>

This is an array reference of key ids in case the mail is encrypted
for several recipients.

=item B<enc_mails>

This is an array reference of the correspondent recipient mail adresses.
Entries may be empty, if gnugpg didn't report the mail address for
a specific key.

=item B<is_signed>

Indicates whether an entity was signed or not.

=item B<sign_ok>

Indicates whether the signature could be verified successfully or not.

=item B<sign_state>

Gives more details about the signature status. May have one of the following
valufes: GOOD, EXP, EXPKEY, REVKEY and BAD. Please refer to gnupg's DETAILS
file for details about the meaning of these values.

=item B<sign_key_id>

The key ID of the sender who signed an entity.

=item B<sign_fingerprint>

Key fingerprint of the sender who signed an entity.

=item B<sign_trust>

Returns how much you trust the signers key. Refer to Mail::GPG->get_key_trust
for a list of known levels and their meaning.

=item B<sign_mail>

The primary mail address of the sender who signed an entity.

=item B<sign_mail_aliases>

A reference to a list of the signer's mail alias addresses.

=item B<gpg_stdout>

This is reference to a scalar containing gpg's STDOUT output.

=item B<gpg_stderr>

This is reference to a scalar containing gpg's STDERR output.

=item B<gpg_rc>

Exit code of the gpg program. Don't rely on this, use the other
attributes to check wether operation was successful resp.
a verification went ok.

=back

=head1 METHODS

There are only two methods, both are for debugging purposes:

=head2 as_string

  $string = $result->as_string ( no_stdout => $no_stdout )

Returns a printable string version of the object.

=over 4

=item no_stdout

If this option is set, gpg's stdout is ommitted in the
string represenation.

=back

=head2 as_short_string

  $short_string = $result->as_short_string;
  
Returns a very short string representation, without any
gpg output, arranged in one line.

=head1 AUTHOR

Joern Reder <joern AT zyn.de>

=head1 COPYRIGHT

Copyright (C) 2004-2006 by Joern Reder, All Rights Reserved.

This library is free software; you can redistribute it
and/or modify it under the same terms as Perl itself.

=head1 SEE ALSO

Mail::GPG, perl(1).

=cut








Makefile.PL
# $Id: Makefile.PL,v 1.5 2011-08-03 08:52:25 joern Exp $

use strict;

use File::Find;
use ExtUtils::MakeMaker;

$| = 1;

#-- Check for patched MIME-tools
eval {
    print "* Checking for patched MIME-tools package... ";
    require MIME::Parser;
    my $parser = MIME::Parser->new;
    $parser->decode_bodies(0);
    print "Ok\n";
};

if ( $@ ) {
    print "Not Ok!\n";
    print "  Please read the README file and apply the required\n";
    print "  MIME-tools patch before installing Mail::GPG.\n";
    print "  You can use Mail::GPG without this patch, but\n";
    print "  Mail::GPG then can't verify all MIME signed messages.\n";
}

#-- check for gpg program
eval {
    print "* Checking for gpg program... ";
    my $out = qx[gpg --version 2>&1 && echo GPGOK];
    die unless $out =~ /GPGOK/;
    print "Ok\n";
};

if ( $@ ) {
    print "Not Ok!\n";
    print "  Please read the README file and first install the\n";
    print "  gpg program in your PATH, or extend your PATH that\n";
    print "  the gpg program can be found. This is needed for\n";
    print "  the regression tests only.\n";
}

WriteMakefile(
    'NAME'              => 'Mail::GPG',
    'VERSION_FROM'      => 'lib/Mail/GPG.pm',
    'PREREQ_PM'         => {
        'MIME::Entity'      => 5.419,
        'MIME::Parser'      => 5.419,
        'Mail::Address'     => 0,
        'MIME::QuotedPrint' => 2.20,
        'GnuPG::Interface'  => 0,
    },
    'EXE_FILES'         => [ 'bin/mgpg-test' ],
    'dist' => {
        COMPRESS => "gzip",
        SUFFIX   => "gz",
        PREOP    => q[./genreadme],
        POSTOP   => q[mkdir -p dist; mv Mail*tar.gz dist/],
    },
);
